File: 1460-number-of-substrings-containing-all-three-characters/number-of-substrings-containing-all-three-characters.rb
# @param {String} s
# @return {Integer}
def number_of_substrings(s)
    i ,j ,ans, temp = 0 , 0 , 0, ""
    counts = Hash.new(0)
    while j < s.size
        temp += s[j]
        counts[s[j]] += 1
        if counts["a"] > 0 && counts["b"] > 0 && counts["c"] > 0
            ans += s.size - j
            while counts["a"] > 0 && counts["b"] > 0 && counts["c"] > 0
                counts[s[i]] -= 1
                if counts["a"] > 0 && counts["b"] > 0 && counts["c"] > 0
                    ans += s.size - j
                end
                i += 1
            end
        end
        j += 1
    end
    ans
end

File: 1460-number-of-substrings-containing-all-three-characters/number-of-substrings-containing-all-three-characters.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        i, j = 0, 0
        valid = 0
        counts = defaultdict(int)
        ans = 0
        while j < len(s):
            counts[s[j]] += 1
            if counts[s[j]] == 1:
                valid += 1
            if valid == 3:
                ans += len(s) - j
            while valid == 3:
                if counts[s[i]] == 1:
                    valid -= 1
                if valid == 3:
                    ans += len(s) - j
                counts[s[i]] -= 1
                i += 1
            j += 1
        return ans